jaxrs/WFLY-10646_EAP7-1034_multivalue_param.adoc

```diff
@@ -0,0 +1,381 @@
+= WFLY-10646 / EAP7-1034 RESTEasy extended support for multivalue @*Param
+:author:            Alessio Soldano
+:email:             asoldano@redhat.com
+:toc:               left
+:icons:             font
+:keywords:          comma,separated,tags
+:idprefix:
+:idseparator:       -
+:issue-base-url:    https://issues.jboss.org/browse
+
+== Overview
+
+This is about supporting multivalue @CookieParam, @HeaderParam, @MatrixParam, @PathParam, @QueryParam, allowing custom conversion from arrays or collections used as parameters to JAX-RS resources' methods.
+
+For parameters and properties annotated with @CookieParam, @HeaderParam, @MatrixParam, @PathParam, or @QueryParam, the JAX-RS specification [https://jcp.org/aboutJava/communityprocess/final/jsr339/index.html] allows conversion as defined in the Javadoc of the corresponding annotation. In general, the following types are supported:
+
+. Types for which a ParamConverter is available via a registered ParamConverterProvider
+. Primitive types
+. Types that have a constructor that accepts a single String argument
+. Types that have a static method named valueOf or fromString with a single String argument that return an instance of the type
+
+With this new feature, RESTEasy is going to also support List<T>, Set<T>, or SortedSet<T>, where T satisfies 3 or 4 above. The way the collections of parameters are expected to be expressed in HTTP messages depends on the particular kind of parameter. In most cases, the notation for collections is based on convention rather than a specification.
```

Comment:
AFAICS, this feature allows users to write their own MultiValuedParamConverter and MultiValuedParamConverterProvider. These classes can allow using multivalued parameters split by `,` character. But users have to create custom logic for list/set/... support and for base collection classes. For example, if I want to use Set<User>, I have to prepared UserParamConverter, or add valueOf/fromString/User(String fromString) to User class. This specific approach needs to be also handled by my own MultiValuedParamConverter/MultiValuedParamConverterProvider, these classes have to know converting approach of User class.

This doesn't work in old WF versions. Anyway, in my PoV, this feature force users to write a lot of feature functionality in deployment. @asoldano / @ronsigal Are there some plans to improve this?

Reply:
I'll let Ron comment on the first question. As for the need of adding user classes in deployment, I don't think that qualifies as a problem here. We're not forcing users to rely on this new feature.
This said, improvements can be evaluate, feel free to provide a proposal.

Reply:
@marekkopecky, there's no need to write anything if you're happy with the conventional syntax. Something like MultiValuedParamConverter is called for only if you want a nonstandard syntax. I see it as a minor but nice extension to JAX-RS. I'm not sure what improvements are called for. One thing that comes to mind is giving a way to declare an alteration to the conventional syntax so that the built-in parsing can be used instead of writing a new MultiValuedParamConverter. I.e., there could be a way to declare '&' as the separator instead of ','. Just a thought.

Reply:
> As for the need of adding user classes in deployment, I don't think that qualifies as a problem here. We're not forcing users to rely on this new feature.
@asoldano  Yes, it is not a functional problem. But:

> built-in parsing can be used instead of writing a new MultiValuedParamConverter
@ronsigal @asoldano In my PoV, this feature would be more user-friendly, if users can use build-in universal MultiValuedParamConverter&&MultiValuedParamConverterProvider from resteasy code-base.